### PR TITLE
Add failsafe to RedisLockExtension to resolve #181

### LINKED ISSUE #181: Infinite lock issue with RedisLockExtension
### What does the bug do?

Produces an infinite lock on resource

### How can it be reproduced?

Aha, I'm struggling to build a reproduction, however it can be described via theory.

### Setup

CacheStack 
Redis layer only
RedisLockExtension used

### What happened?

We started using background refresh in production, and we found we were losing MySQL connections - they effectively got taken from the pool and never returned. This process happens slowly - a couple every few 100k requests.

Initially we turned off background refreshing, but issue was still present. We then just used Get/Set functions (as we did previously) and the issue was not there. Finally we removed the `RedisLockExtension` without background refreshing, and the issue was not there, and then added background refreshing back in, and issue still not there. This effectively points the finger at `RedisLockExtension`. We have been stable in production for 48 hours without it (no lost connections from the pool), and prior to this we were losing connections every hour or 2.

Whats going on?

When we build a context via our IoC container, we inject a DB session - under the hood this is grabbing a MySQL connection. Also when most web requests run, we also grab a DB session (and therefore MySQL connection) for the request.

Infinite lock issue

It's taken me a while to spot:

1) Try to get a lock - https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L73

2) If we can't get a lock, we wait on the completion source: https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L101

3) Once the owner of the lock has refreshed a value, it publishes that via redis: https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L80

4) The subs try to clear down that completion source by setting the result: https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L50

But what happens if there is a disruption to the Redis connection, and the sub never receives the lock release? Well, in this case you have an infinite lock, and there is no way to release it.

Although a similar process is used in the main CacheStack - https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower/CacheStack.cs#L313 - We aren't relying on an external dependency to transmit a message to make sure we unlock, instead it happens in the thread context that performed the update and this issue cannot occur.

Thoughts:

* We could use a timeout on the completion source? Make sure the lock always releases even if no sub received.
* I wonder if a busy wait approach on LockTakeAsync/LockReleaseAsync would be better, rather than relying on pub/sub?

### PR DESCRIPTION
**Pull request by mgoodfellow**
PR idea for resolving #181

### CODE REVIEW

File: src/CacheTower.Extensions.Redis/RedisLockExtension.cs [hunk 1]
```diff
@@ -87,7 +88,13 @@ public async ValueTask<CacheEntry<T>> WithRefreshAsync<T>(string cacheKey, Func<
 			}
 			else
 			{
-				var completionSource = LockedOnKeyRefresh.GetOrAdd(cacheKey, key => new TaskCompletionSource<bool>());
+				var completionSource = LockedOnKeyRefresh.GetOrAdd(cacheKey, key =>
+				{
+					var tcs = new TaskCompletionSource<bool>();
+					var cts = new CancellationTokenSource(Options.LockTimeout);
+					cts.Token.Register(() => tcs.TrySetCanceled(), useSynchronizationContext: false);
```
Comment (Turnerj) [inline src/CacheTower.Extensions.Redis/RedisLockExtension.cs:95]:
```suggestion
					cts.Token.Register(tcs => ((TaskCompletionSource<bool>)tcs).TrySetCanceled(), tcs, useSynchronizationContext: false);
```

This helps avoid some allocations by avoiding a closure. There is technically a second optimization by switching to `TaskCompletionSource` (the non-generic one) too but that is out-of-scope of this issue.

**Review by Turnerj — CHANGES_REQUESTED** [review]

File: src/CacheTower.Extensions.Redis/RedisLockExtension.cs (hunk 1 above)
Comment (mgoodfellow) [reply src/CacheTower.Extensions.Redis/RedisLockExtension.cs]: Yeah, I did notice we don't use the return value of that. I figured a task cancelled thrown up the stack was the only outcome that makes sense in this case. Ah those tricky closure allocations! Thanks for the suggestion, added to the PR

**Review by Turnerj — APPROVED** [review]
Looks good to me!

**Comment by Turnerj** [comment]
Oh, if you were curious, this is what I got benchmarking etc to find the amount the closure affected the code:

|   Method |      Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------- |----------:|----------:|----------:|------:|------:|------:|----------:|
| Before fix |  10.03 ns |  0.617 ns |  0.660 ns |     - |     - |     - |      96 B |
| Your original commit | 646.30 ns | 84.853 ns | 97.717 ns |     - |     - |     - |     569 B |
| Without additional closure | 522.62 ns | 51.334 ns | 54.926 ns |     - |     - |     - |     481 B |
| Non-generic TCS | 499.49 ns | 17.859 ns | 18.340 ns |     - |     - |     - |     472 B |

**Comment by mgoodfellow** [comment]
That is interesting, thanks!

Of course, this isn't really a "hot path" as this code path can only run maximum of once if the lock is already taken. Other code paths won't create the tcs or cts, instead waiting on the one previously created.

The impact of all these closures add up though, I definitely use them liberally as I am a fan of the syntax. I've never really had to focus on raw performance code, so this is all interesting stuff to me!

Thanks for the merge, I'm going to shipping to production in next day or 2 and will monitor!

**Comment by Turnerj** [comment]
Yeah, definitely not a hot path so I definitely wouldn't bother doing anything too exotic but still nice getting a few small wins.

More important than performance though is that this bug is fixed. Having buggy but fast code isn't particularly helpful.

**Pull request merged**